File: internal/observability/tracer_filter.go
package observability

import (
	"context"

	"go.opentelemetry.io/otel/trace"
	"go.opentelemetry.io/otel/trace/embedded"
	nooptrace "go.opentelemetry.io/otel/trace/noop"
)

// filteringTracerProvider wraps a real TracerProvider and suppresses
// hot-path spans to keep trace volume manageable. Entire tracer names
// can be suppressed (returning noop tracers), and individual span names
// can be suppressed within otherwise-active tracers.
type filteringTracerProvider struct {
	embedded.TracerProvider

	delegate          trace.TracerProvider
	noop              trace.TracerProvider
	suppressedTracers map[string]bool
	suppressedSpans   map[string]bool
}

// NewFilteringTracerProvider wraps delegate so that hot-path spans are
// replaced with no-op spans. This drops per-commit/per-file/per-git-op
// spans while preserving structural pipeline spans.
func NewFilteringTracerProvider(delegate trace.TracerProvider) trace.TracerProvider {
	return &filteringTracerProvider{
		delegate: delegate,
		noop:     nooptrace.NewTracerProvider(),
		suppressedTracers: map[string]bool{
			"codefang.gitlib": true,
			"codefang.uast":   true,
		},
		suppressedSpans: map[string]bool{
			"codefang.analyzer.consume": true,
		},
	}
}

// Tracer returns a tracer for the given name, suppressing hot-path tracers.
func (f *filteringTracerProvider) Tracer(name string, opts ...trace.TracerOption) trace.Tracer {
	if f.suppressedTracers[name] {
		return f.noop.Tracer(name, opts...)
	}

	actual := f.delegate.Tracer(name, opts...)

	if len(f.suppressedSpans) > 0 {
		return &filteringTracer{
			delegate: actual,
			noop:     f.noop.Tracer(name, opts...),
			suppress: f.suppressedSpans,
		}
	}

	return actual
}

// filteringTracer wraps a real Tracer and returns noop spans for
// suppressed span names while delegating everything else.
type filteringTracer struct {
	embedded.Tracer

	delegate trace.Tracer
	noop     trace.Tracer
	suppress map[string]bool
}

// Start creates a span, returning a noop span for suppressed names.
func (f *filteringTracer) Start(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
	if f.suppress[name] {
		return f.noop.Start(ctx, name, opts...)
	}

	return f.delegate.Start(ctx, name, opts...)
}

File: rust/crates/cf-observability/src/tracer_filter.rs
//! Hot-path tracer/span suppression.
//!
//! Port of `internal/observability/tracer_filter.go`. Wraps a real
//! `TracerProvider` so per-commit/per-file/per-git-op spans are replaced with
//! no-op spans while structural pipeline spans survive.
//!
//! # Decision logic vs SDK wiring
//!
//! Go composes this as a `trace.TracerProvider` whose `Tracer(name)` returns a
//! no-op tracer for suppressed tracer names, and otherwise wraps the real tracer
//! so suppressed span names start no-op spans. The two decision points are
//! [`is_tracer_suppressed`] and [`is_span_suppressed`], which carry the exact
//! suppression sets from the Go source. [`FilteringTracerProvider`] composes
//! them over any [`opentelemetry::trace::TracerProvider`]; the binding test for
//! the suppression sets is on the pure predicates (deterministic), while the
//! provider wrapper reproduces the dispatch behavior.

use std::collections::HashSet;

/// Tracer names whose spans are entirely suppressed (Go `suppressedTracers`).
pub const SUPPRESSED_TRACERS: &[&str] = &["codefang.gitlib", "codefang.uast"];

/// Span names suppressed even within otherwise-active tracers
/// (Go `suppressedSpans`).
pub const SUPPRESSED_SPANS: &[&str] = &["codefang.analyzer.consume"];

/// Returns true if every span from tracer `name` should be suppressed.
///
/// Port of the `suppressedTracers[name]` lookup.
#[must_use]
pub fn is_tracer_suppressed(name: &str) -> bool {
    SUPPRESSED_TRACERS.contains(&name)
}

/// Returns true if a span named `name` should be suppressed.
///
/// Port of the `suppressedSpans[name]` lookup.
#[must_use]
pub fn is_span_suppressed(name: &str) -> bool {
    SUPPRESSED_SPANS.contains(&name)
}

/// Wraps a delegate [`TracerProvider`](opentelemetry::trace::TracerProvider) so
/// hot-path spans become no-op spans (Go `filteringTracerProvider`).
///
/// This is parameterized over the delegate provider type so it can wrap either
/// the SDK provider or a no-op provider, matching Go's `NewFilteringTracerProvider`.
pub struct FilteringTracerProvider<P> {
    delegate: P,
    suppressed_tracers: HashSet<&'static str>,
    suppressed_spans: HashSet<&'static str>,
}

impl<P> FilteringTracerProvider<P> {
    /// Wraps `delegate` with the default Codefang suppression sets
    /// (Go `NewFilteringTracerProvider`).
    #[must_use]
    pub fn new(delegate: P) -> Self {
        FilteringTracerProvider {
            delegate,
            suppressed_tracers: SUPPRESSED_TRACERS.iter().copied().collect(),
            suppressed_spans: SUPPRESSED_SPANS.iter().copied().collect(),
        }
    }

    /// Returns whether spans from tracer `name` are entirely suppressed.
    #[must_use]
    pub fn tracer_suppressed(&self, name: &str) -> bool {
        self.suppressed_tracers.contains(name)
    }

    /// Returns whether span `name` is suppressed within active tracers.
    #[must_use]
    pub fn span_suppressed(&self, name: &str) -> bool {
        self.suppressed_spans.contains(name)
    }

    /// Borrows the wrapped delegate provider.
    pub fn delegate(&self) -> &P {
        &self.delegate
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Port of Go `TestFilteringProvider_SuppressedTracer`.
    #[test]
    fn suppressed_tracer() {
        assert!(is_tracer_suppressed("codefang.gitlib"));
    }

    /// Port of Go `TestFilteringProvider_UASTParseSuppressed`.
    #[test]
    fn uast_tracer_suppressed() {
        assert!(is_tracer_suppressed("codefang.uast"));
    }

    /// Port of Go `TestFilteringProvider_SuppressedSpan` (hot-path span dropped,
    /// structural span passes).
    #[test]
    fn suppressed_span_vs_structural() {
        assert!(is_span_suppressed("codefang.analyzer.consume"));
        assert!(!is_span_suppressed("codefang.runner.run"));
    }

    /// Port of Go `TestFilteringProvider_PassThrough` (root tracer not suppressed).
    #[test]
    fn pass_through_root_tracer() {
        assert!(!is_tracer_suppressed("codefang"));
        assert!(!is_span_suppressed("codefang.some_operation"));
    }

    #[test]
    fn wrapper_exposes_decisions() {
        let fp = FilteringTracerProvider::new(());
        assert!(fp.tracer_suppressed("codefang.gitlib"));
        assert!(!fp.tracer_suppressed("codefang.framework"));
        assert!(fp.span_suppressed("codefang.analyzer.consume"));
        assert!(!fp.span_suppressed("codefang.runner.run"));
    }
}
